File: leet-code/maxDepth.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * 后序遍历
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function (root) {
  return oderTree(root)
}

function oderTree (node) {
  if (node == null) {
    return 0
  }
  return Math.max(oderTree(node.left), oderTree(node.right)) + 1
}

/**
 * 层序遍历
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function (root) {
  if (root == null) {
    return 0
  }
  return sequenceOrderTree(root)
}

function sequenceOrderTree (node) {
  let count = 0
  let queue = []
  queue.push(node)
  while (queue.length > 0) {
    let levelSize = queue.length
    for (let i = 0; i < levelSize; i++) {
      let node = queue.shift()
      if (node.left != null) {
        queue.push(node.left)
      }
      if (node.right != null) {
        queue.push(node.right)
      }
    }
    count++
  }
  return count
}

console.log(maxDepth({
  val: 1,
  left: {
    val: 2,
    left: {
      val: 4,
      left: null,
      right: null
    },
    right: {
      val: 5,
      left: null,
      right: null
    }
  },
  right: {
    val: 3,
    left: {
      val: 6,
      left: null,
      right: null
    },
    right: {
      val: 7,
      left: null,
      right: null
    }
  }
}))

File: leet-code/maxDepth.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// 后序遍历解法
func maxDepth(root *TreeNode) int {
	return backOrderTree(root)
}

func backOrderTree(root *TreeNode) int {
	if root == nil {
		return 0
	}
	var max func(x, y int) int
	max = func(x, y int) int {
		if x > y {
			return x
		}
		return y
	}
	return max(backOrderTree(root.Left), backOrderTree(root.Right)) + 1
}

// 层序遍历解法
func maxDepth1(root *TreeNode) int {
	return sequenceOrderTree(root)
}

func sequenceOrderTree(root *TreeNode) int {
	if root == nil {
		return 0
	}
	queue := []*TreeNode{root}
	count := 0
	for len(queue) > 0 {
		levelSize := len(queue)
		for i := 0; i < levelSize; i++ {
			node := queue[i]
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}
		count++
		queue = queue[levelSize:]
	}
	return count
}
